maas/resource_maas_boot_source.go

```diff
@@ -1,12 +1,126 @@
+
+		Schema: map[string]*schema.Schema{
+			"created": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The creation time of the boot source.",
+			},
+			"keyring_data": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The data on the keyring for the boot source.",
+			},
+			"keyring_filename": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The filename on the keyring for the boot source.",
+			},
+			"updated": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The time of most recent update of the boot source.",
+			},
+			"url": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The URL of the boot source.",
+			},
+		},
+	}
+}
+
+func resourceBootSourceCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*client.Client)
+
+	bootsourceParams := entity.BootSourceParams{
+		KeyringData:     d.Get("keyring_data").(string),
+		KeyringFilename: d.Get("keyring_filename").(string),
+		URL:             d.Get("url").(string),
+	}
+
+	bootsource, err := client.BootSources.Create(&bootsourceParams)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId(fmt.Sprintf("%v", bootsource.ID))
+
+	return resourceBootSourceRead(ctx, d, meta)
+}
+
+func resourceBootSourceRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*client.Client)
+
+	bootsource, err := getBootSource(client)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId(fmt.Sprintf("%v", bootsource.ID))
+
+	tfState := map[string]interface{}{
+		"created":          bootsource.Created,
+		"keyring_data":     bootsource.KeyringData,
+		"keyring_filename": bootsource.KeyringFilename,
+		"updated":          bootsource.Updated,
+		"url":              bootsource.URL,
+	}
+	if err := setTerraformState(d, tfState); err != nil {
+		return diag.FromErr(err)
+	}
+
+	return nil
+}
+
+func resourceBootSourceUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*client.Client)
+
+	bootsourceParams := entity.BootSourceParams{
+		KeyringData:     d.Get("keyring_data").(string),
+		KeyringFilename: d.Get("keyring_filename").(string),
+		URL:             d.Get("url").(string),
+	}
+	id, err := strconv.Atoi(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	bootsource, err := client.BootSource.Update(id, &bootsourceParams)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId(fmt.Sprintf("%v", bootsource.ID))
+
+	return resourceBootSourceRead(ctx, d, meta)
+}
+
+func resourceBootSourceDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*client.Client)
+	id, err := strconv.Atoi(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	return diag.FromErr(client.BootSource.Delete(id))
```

Comment:
Based on the assumption that there is always one boot source that cannot be replaced, etc. the delete function should just revert the URL to point to the default URL that MAAS is setting, that is the URL of stable images